rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,195 @@
+
+3. It is confusing (IMO) that passing a render function as `children` would override the _whole slot_, and not just its children.
+
+And a less important issue that's addressed in Appendix A of this proposal:
+
+4. The order of the arguments is `(Component, props)`, but one of the primary use cases is to replace the given slot component with something else, in which case you ignore the first argument. You'd rarely/never(?) want to ignore the second argument `props`.
+
+## Proposal
+
+The core proposal in this RFC is to use the `as` prop as the render function instead of `children`.
+
+```jsx
+<>
+  {/* Existing behavior: unintentionally overwrites default children of the button slot. */}
+  <AccordionHeader button={{ children: (Component, props) => <Component {...props} /> }} />
+
+  {/* New behavior, use 'as' instead, and `props` includes the slot's default `children`. */}
+  <AccordionHeader button={{ as: (Component, props) => <Component {...props} /> }} />
+</>
+```
+
+You can still pass in a string to the `as` prop (e.g. `as="a"` to render an `<a>` tag), and it would work as it does today:
+
+- If the slot's type is an intrinsic element like `'button'`, it replaces that element with the given one.
+- If the slot's type is a component like `Button`, it forwards the `as` prop to the component.
+
+> 🤔 What if you want to provide a render function _and_ override the component's `as` prop? Write a render function that adds your own `as` and spreads props:
+>
+> ```jsx
+> <Checkbox label={{ children: 'Label text', as: (_ignored, props) => <MyLabel as="span" {...props} /> }} />
+> ```
+
+### Pros
+
+1. 👍 Allows us to pass the slot's default `children` to the render function, because it doesn't require overriding the default children.
+
+   ```jsx
+   <AvatarGroup
+     overflowLabel={{
+       as: (Component, props) => (
+         <Component {...props}>
+           {...props.children}
+           <span>extra text</span>
+         </Component>
+       ),
+     }}
+   />
+   ```
+
+2. 👍 Allows the component to use children as a function if it needs to, without interfering with the ability for the user to write a slot render function for the root slot:
+
+   ```jsx
+   <Field as={(_, props) => <MyCustomRoot foo="bar" {...props} />}>
+     {inputProps => (
+       <div>
+         <input {...inputProps} />
+       </div>
+     )}
+   </Field>
+   ```
+
+3. 👍 The slot's `children` prop is no longer "special": it's treated the same as any other prop. This reduces the slot API to a single "special" prop: `as`.
+
+4. 👍 Opinion: it's easier to understand the API. The `as` prop implies that the render function is rendering the whole slot. Whereas `children` implies that it's rendering just the children, which is not the case.
+
+### Cons
+
+1. 👎 Overloads the `as` prop, which may run into issues in implementation, if there are any other uses of `as` as a function. See "Appendix B" for an alternative proposal that uses a prop `render` instead of `as`.
```

Comment:
I'm not sure if this is such a downside, after all we're extending `as` prop with a render function possibility, which kind of looks like a generic alternative to `as`